pkg/api/policy/policy.go

```diff
@@ -0,0 +1,171 @@
+	if err := e.LoadRules(false); err != nil {
+		return false, err
+	}
+
+	toRule, ok := e.Rules[rule]
+	if !ok {
+		err := fmt.Errorf("Rule [%s] does not exist", rule)
+		return false, err
+	}
+	return check(toRule, target, cred, *e, ""), nil
+}
+
+func (e *Enforcer) Authorize(rule string, target map[string]string, cred map[string]interface{}) (bool, error) {
+	return e.Enforce(rule, target, cred)
+}
+
+func (e *Enforcer) LoadRules(forcedReload bool) error {
+	path := config.CONF.OsdsLet.PolicyPath
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	// Load all policy files that in the specified path
+	if fileInfo.IsDir() {
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			return err
+		}
+		for _, f := range files {
+			if !f.IsDir() && strings.HasSuffix(f.Name(), ".json") {
+				err := e.LoadPolicyFile(path, forcedReload, false)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	} else {
+		return e.LoadPolicyFile(path, forcedReload, e.OverWrite)
+	}
+}
+
+func (e *Enforcer) UpdateRules(rules map[string]BaseCheck) {
+	if e.Rules == nil {
+		e.Rules = make(map[string]BaseCheck)
+	}
+	for k, c := range rules {
+		e.Rules[k] = c
+	}
+}
+
+func (e *Enforcer) LoadPolicyFile(path string, forcedReload bool, overWrite bool) error {
+	// if rules is already set or user doesn't want to force reload, return it.
+	if e.Rules != nil && !forcedReload {
+		return nil
+	}
+
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		msg := fmt.Sprintf("Read policy file (%s) failed, reason:(%v)", path, err)
+		log.Error(msg)
+		return fmt.Errorf(msg)
+	}
+
+	rules, err := NewRules(data, e.DefaultRule)
+	if err != nil {
+		return err
+	}
+
+	if overWrite {
+		e.Rules = rules.Rules
+	} else {
+		e.UpdateRules(rules.Rules)
+	}
+	return nil
+}
+
+func NewRules(data []byte, defaultRule string) (*Rules, error) {
+	rules := &Rules{}
+	err := rules.Load(data, defaultRule)
+	return rules, err
+}
+
+type Rules struct {
+	Rules       map[string]BaseCheck
+	DefaultRule string
+}
+
+func (r *Rules) Load(data []byte, defaultRule string) error {
+	rulesMap := map[string]string{}
+	err := json.Unmarshal(data, &rulesMap)
+	if err != nil {
+		err := fmt.Errorf("Json unmarshal failed:", err)
+		log.Errorf(err.Error())
+		return err
+	}
+
+	if r.Rules == nil {
+		r.Rules = make(map[string]BaseCheck)
+	}
+	for k, v := range rulesMap {
+		r.Rules[k] = parseRule(v)
+	}
+	r.DefaultRule = defaultRule
+	return nil
+}
+
+func (r *Rules) String() string {
+	b, _ := json.MarshalIndent(r.Rules, "", "  ")
+	return string(b)
+}
+
+func Authorize(ctx *filter.Context, action string) bool {
+	credentials := ctx.ToPolicyValue()
+	target := map[string]string{
+		"project_id": ctx.ProjectId,
+		"user_id":    ctx.UserId,
+	}
+	log.V(8).Infof("Action: %v", action)
+	log.V(8).Infof("Target: %v", target)
+	log.V(8).Infof("Credentials: %v", credentials)
+	ok, err := enforcer.Authorize(action, target, credentials)
+	if err != nil {
```

Comment:
What do you mean to return ok all the time? It should be handled when some errors occur.

Reply:
It just a variable name, it only stand for if it authorize success. one situation is that authorize failed, but there is not any error happen, it only represents the user without permission.

Reply:
Got it